src/core/qgsvectordataprovider.h

```diff
@@ -160,6 +160,25 @@ class CORE_EXPORT QgsVectorDataProvider : public QgsDataProvider, public QgsFeat
      */
     long featureCount() const override = 0;
 
+    /**
+     * Returns true if the layer contains at least one feature.
+     *
+     * \since QGIS 3.2
+     */
+    virtual bool empty() const;
+
+    /**
+     * Will always return FeatureAvailability::FeaturesAvailable or
+     * FeatureAvailability::FeaturesMaybeAvailable.
+     *
+     * Calls empty() internally. Providers should override empty instead
+     * if they provide an optimized version of this call.
+     *
+     * \since QGIS 3.2
+     * \see empty()
+     */
+    virtual QgsFeatureSource::FeatureAvailability hasFeatures() const override;
```

Comment:
I think it's redundant since the override keyword was introduced. And now tools like the clang tidy modernize checks automatically strip it out too.

Reply:
One could claim that the `override` keyword itself is redundant information too, still I'd never argue to remove that for the technical safety it adds. I'm happy to also keep some "clutter" around for visual clarity (like typenames instead of auto). Anyway, let's go for the clang tidy standard :stuck_out_tongue_closed_eyes: